### PR TITLE
upgrading InfogainLoss layer 

### PR DESCRIPTION
**Pull request by shaibagon**
Making InfogainLoss layer more similar to SoftmaxWithLoss in term of functionality:
(1) incorporating Softmax layer to make the gradient computation robust, much like SoftmaxWithLoss layer (see: [this SO thread](http://stackoverflow.com/a/34917052/1714410) for more information). 
(2) supporting loss along axis (see [this SO thread for motivation](http://stackoverflow.com/q/36115026/1714410)).

Updated `test_infogain_loss_layer.cpp` accordingly.

This PR is self-contained: it only affects infogain loss layer.

### CODE REVIEW

**Comment by gburachas** [comment]
Thanks for the valuable contribution! Would you mind providing the the snippet example of protobuf code showing how to define the layer.
Thanks-
GTB

**Comment by shaibagon** [comment]
@gburachas the upgraded layer "look and feel" exactly like the old one, apart from the fact that you no longer need to explicitly pass the input through a `"Softmax"` layer.
You can find more information at stackoverflow, [here](http://stackoverflow.com/q/27632440/1714410) and in some linked threads there.

**Comment by gburachas** [comment]
Thanks for the response! Any suggestions regarding how to define the infogain matrix H for N=1000 categories with unbalanced categories? Instance counts per category in my database vary by x4 (between 300 and 1200). Using 1-f(i) for diagonal entries results in very similar weightings for all categories. I'm considering something along the lines 1-g(i), where g(i) = ((count_i - min_count)/total_count)^(1/4) or so (min_count is the minimum count for a category in the db).
Thanks-
GTB

**Comment by shaibagon** [comment]
@gburachas  why not put in diagonal `1/|i|` where `|i|` is the number of samples marked as label `i`? if sample `x` that should be labeled `i`, got high probability for `j != i`, then the loss is `-1/|i| * log(p(x))` where `-log(p(x))` is high (since `x` got low probability to be labeled as `i`).

**Comment by gburachas** [comment]
@shaibagon good idea, thanks!

**Comment by gburachas** [comment]
one more question: can I take the randomly oscillating loss and accuracy value (around random guessing)  while training for a sign of numeric instability? I'm running a training session with the infogain matrix as discussed above, but decided to give a shot at the original layers, softmax+infogain_loss. After ~12K iterations the loss and accuracy are not showing signs of improvement..

**Comment by shaibagon** [comment]
@gburachas I am starting to feel this thread is not the right place for this discussion. Can you ask this question on stackoverflow? tag it with "caffe".

**Comment by shaibagon** [comment]
related SO [thread](http://stackoverflow.com/q/37491507/1714410).

**Comment by moi90** [comment]
I think `InfogainLossLayer` has a right to exist as it is. IMHO, you should rename your contribution to `SoftmaxWithInfogainLoss` analogous to `SoftmaxWithLoss` or `SigmoidCrossEntropyLossLayer`.

**Comment by shaibagon** [comment]
@moi90 since the computation of infogain is assuming class probabilities as inputs (and it is quite needed bcause of the `log`), can you describe a use case where infogain is used without softmax layer preceding it?

**Comment by moi90** [comment]
There are more ways to create a discrete probability distribution than softmax.

**Comment by shaibagon** [comment]
@stormkingz  Thank you for approving this PR!

**Comment by shaibagon** [comment]
@stormkingz after approving this PR, when is it expected to be merged to master?

Thank you!

**Comment by shaibagon** [comment]
Another example of using pixel-wise `"InfogainLoss"` to tackle class imbalance in semantic-segmentation task can be found [at this SO thread](http://stackoverflow.com/q/41670022/1714410).

**Comment by shaibagon** [comment]
Yet another user waiting for this change [here](http://stackoverflow.com/q/42344936/1714410)...

**Comment by beniz** [comment]
@shaibagon FYI, at deepdetect we are maintaining a version of Caffe with Softmax, see https://github.com/beniz/caffe/commit/f36cf2b163cc424317295ccc4f6a70852351a42c
This is useful is some cases to mitigate false positives for instance. Possibly it may be of some use for you and others.
Repo is https://github.com/beniz/caffe

That being said, it'd be great to have it in vanilla Caffe I reckon.

**Comment by shaibagon** [comment]
@beniz thank you for the comment and contributions. I hope the Gods-of-Caffe will find the time to accept this PR...

**Comment by shelhamer** [comment]
Thanks Shai for the improved loss layer!

**Comment by zllljf** [comment]
hello 
i compile caffe with this bug:
`CXX src/caffe/layers/infogain_loss_layer.cpp
src/caffe/layers/infogain_loss_layer.cpp: In instantiation of ‘void caffe::InfogainLossLayer<Dtype>::LayerSetUp(const std::vector<caffe::Blob<Dtype>*>&, const std::vector<caffe::Blob<Dtype>*>&) [with Dtype = float]’:
src/caffe/layers/infogain_loss_layer.cpp:221:1:   required from here
src/caffe/layers/infogain_loss_layer.cpp:18:3: error: ‘const class caffe::InfogainLossParameter’ has no member named ‘axis’
   softmax_param->set_axis(this->layer_param_.infogain_loss_param().axis());
   ^
src/caffe/layers/infogain_loss_layer.cpp: In instantiation of ‘void caffe::InfogainLossLayer<Dtype>::Reshape(const std::vector<caffe::Blob<Dtype>*>&, const std::vector<caffe::Blob<Dtype>*>&) [with Dtype = float]’:
src/caffe/layers/infogain_loss_layer.cpp:221:1:   required from here
src/caffe/layers/infogain_loss_layer.cpp:55:18: error: ‘const class caffe::InfogainLossParameter’ has no member named ‘axis’
   infogain_axis_ =
                  ^
src/caffe/layers/infogain_loss_layer.cpp: In instantiation of ‘void caffe::InfogainLossLayer<Dtype>::LayerSetUp(const std::vector<caffe::Blob<Dtype>*>&, const std::vector<caffe::Blob<Dtype>*>&) [with Dtype = double]’:
src/caffe/layers/infogain_loss_layer.cpp:221:1:   required from here
src/caffe/layers/infogain_loss_layer.cpp:18:3: error: ‘const class caffe::InfogainLossParameter’ has no member named ‘axis’
   softmax_param->set_axis(this->layer_param_.infogain_loss_param().axis());
   ^
src/caffe/layers/infogain_loss_layer.cpp: In instantiation of ‘void caffe::InfogainLossLayer<Dtype>::Reshape(const std::vector<caffe::Blob<Dtype>*>&, const std::vector<caffe::Blob<Dtype>*>&) [with Dtype = double]’:
src/caffe/layers/infogain_loss_layer.cpp:221:1:   required from here
src/caffe/layers/infogain_loss_layer.cpp:55:18: error: ‘const class caffe::InfogainLossParameter’ has no member named ‘axis’
   infogain_axis_ =
                  ^
Makefile:581: recipe for target '.build_release/src/caffe/layers/infogain_loss_layer.o' failed
make: *** [.build_release/src/caffe/layers/infogain_loss_layer.o] Error 1
`
i change the caffe.proto with adding a new loss layer, how can i deal with it?
thank you for your reply.
